### PR TITLE
Introduce DefaultValue type

### LINKED ISSUE #3910: Refactor Configuration off `String` type for default value.
Currently, we are formatting the default value into a free-form `String` in `Configuration`. This is bypassing the typing system and couples the `ConfigurationCollector` with `ConfigPrinter` implicitly.

_Originally posted by @chao2zhang in https://github.com/detekt/detekt/pull/3827#discussion_r643217935_

### PR DESCRIPTION
**Pull request by marschwar**
This PR resolves #3910. The change set is greater than I anticipated but I think the collection and the generation are better separated now and there is less string manipulation magic involved.

### CODE REVIEW

**Comment by marschwar** [comment]
This will need to be reworked after #3964 is merged.

**Comment by cortinico** [comment]
> This will need to be reworked after #3964 is merged.

Can we follow up on this?

**Comment by marschwar** [comment]
Sorry, I have been "offline" for a while. Will try to update soon.

**Comment by marschwar** [comment]
Any thoughts on this? Does this make sense to you?

**Review by BraisGabin — APPROVED** [review]
Sorry for the delay in the review. I didn't saw that this was ready to review. LGTM

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+private data class BooleanDefault(private val defaultValue: Boolean) : DefaultValue {
+    override fun getAsPlainString(): String = defaultValue.toString()
+}
+
+private data class IntegerDefault(private val defaultValue: Int) : DefaultValue {
+    override fun getAsPlainString(): String = defaultValue.toString()
+}
+
+private data class ListDefault(private val defaultValue: List<String>) : DefaultValue {
```
Comment (chao2zhang) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt:31]: I would call this `ListString` instead of `List`

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/ConfigurationCollector.kt [hunk 2]
```diff
@@ -6,6 +6,10 @@ import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.C
 import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.ConfigWithFallbackSupport.FALLBACK_DELEGATE_NAME
 import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.ConfigWithFallbackSupport.checkUsingInvalidFallbackReference
 import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.ConfigWithFallbackSupport.isFallbackConfigDelegate
+import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.DefaultValueSupport.getAndroidDefaultValue
+import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.DefaultValueSupport.getDefaultValue
+import io.gitlab.arturbosch.detekt.generator.collection.ConfigurationCollector.DefaultValueSupport.toDefaultValueIfLiteral
+import io.gitlab.arturbosch.detekt.generator.collection.DefaultValue.Companion.of
```
Comment (chao2zhang) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/ConfigurationCollector.kt:12]: Hygiene: What do you think about not importing this so that instead of the confusing `of()`, we would use `DefaultValue.of()`

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package io.gitlab.arturbosch.detekt.generator.collection
+
+sealed interface DefaultValue {
+    fun isNonEmptyList(): Boolean = false
+    fun getAsList(): List<String> = error("default value is not a list")
+    fun getAsPlainString(): String
```
Comment (chao2zhang) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt:6]: Could we default this to `defaultValue.toString()`?

File: detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt [hunk 4]
```diff
@@ -195,12 +196,12 @@ object RuleCollectorSpec : Spek({
                          */
                         class SomeRandomClass() : Rule {
                             @Configuration("description")
-                            private val config: Int by config(99)
+                            private val config: Int by config(1_999_000)
```
Comment (chao2zhang) [inline detekt-generator/src/test/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleCollectorSpec.kt:199]: Nice test of the edge case.

**Review by chao2zhang — APPROVED** [review]

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt (hunk 1 above)
Comment (marschwar) [reply detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/DefaultValue.kt:31]: Do you mean `ListDefault` :arrow_right: `ListStringDefault`? If so, I like the idea, but I would prefer `StringListDefault`.
